File: Grade_2/2-1/DS/study/0414/binary_search.c
#include <stdio.h>
#define FAIL -1

int binary_search(int arr[], int first, int len, int find_num){
    int mid;
    int last = len - 1;

    while (first <= last) {
        mid = (first + last) / 2;
        printf("first: %d, end: %d, mid: %d\n", first, last, mid);

        if (find_num == arr[mid])
            return mid;
        else if (find_num > arr[mid])
            first = mid + 1;
        else
            last = mid - 1;
    }

    return FAIL;
}

int main () {
    int arr[10] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    int num;
    int first = 0;
    int len = sizeof(arr) / sizeof(int) ;

    printf("숫자 입력: ");
    scanf("%d", &num);

    printf("%d\n", binary_search(arr, first, len, num));
}

File: Grade_2/2-1/DS/study/0414/binary_search.py
import sys

def binary_search(arr, first, last, find_num):
    mid = (first + last) // 2

    print(f'idx: first: {first}, mid: {mid}, last: {last}')
    print(f'num: first: {arr[first]}, mid: {arr[mid]}, last: {arr[last]}\n')

    if find_num == arr[mid]:
        return find_num
    elif find_num > arr[mid]:
        return binary_search(arr, mid + 1, last, find_num)
    else:
        return binary_search(arr, first, mid - 1, find_num)

arr = [i for i in range(1, 10)]
first = 0
length = len(arr)

find_num = int(sys.stdin.readline().rstrip())
print(f'result = {binary_search(arr, first, length - 1, find_num)}')